cli/deno_error.rs

```diff
@@ -1,6 +1,18 @@
 // Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
-use crate::diagnostics::Diagnostic;
-use crate::fmt_errors::JSError;
+
+/// This module implements error serialization; it
+/// allows to serialize Rust errors to be sent to JS runtime.
+///
+/// Currently it is deeply intertwined with `ErrBox` which is
+/// not optimal since not every ErrBox can be "JS runtime error";
+/// eg. there's no way to throw JSError/Diagnostic from within JS runtime
```

Comment:
There are many types of errors in Deno:
- ErrBox: a generic boxed object. This is the super type of all errors handled in Rust.
- JSError: exceptions thrown from V8 into Rust. Usually a user exception. These are basically a big JSON structure which holds information about line numbers. We use this to pretty-print stack traces. These are never passed back into the runtime.
- DenoError: these are errors that happen during ops, which are passed back into the runtime, where an exception object is created and thrown. DenoErrors have an integer code associated with them - access this via the kind() method.
- Diagnostic: these are errors that originate in TypeScript's compiler. They're similar to JSError, in that they have line numbers. But Diagnostics are compile-time type errors, whereas JSErrors are runtime exceptions.